url146c819f25_jlink_Parse4SJM_tgz-pJ8-sjm_examples_book_robot_ShowRobotRefactoredJ8/src/sjm/examples/book/sling/SlingMediator.java

```diff
@@ -0,0 +1,240 @@
+				throw new RuntimeException("Expected Command and found " + o.getClass());
+			}
+		}
+		return cs;
+	}
+
+	/*
+	 * Parse the user's program, if it has changed, into a command. Execute the
+	 * command in a separate thread to allow halting. Plot whatever functions
+	 * the program specifies.
+	 */
+	protected void go() {
+		try {
+			String thisProgram = programArea.getText();
+			if (!thisProgram.equals(lastProgram)) {
+				lastProgram = thisProgram;
+				command = parse(thisProgram);
+				setComputing(true);
+				computeThread = new Thread(this);
+				computeThread.start();
+			} else {
+				plotPanel.setPlot(target.getRenderables());
+			}
+		} catch (Exception e) {
+			String text = e.toString();
+			if (e.getMessage() != null) {
+				text = e.getMessage();
+			}
+			messageArea.append(text + "\n");
+			return;
+		}
+	}
+
+	/**
+	 * Initialize this mediator with components from the a Sling IDE.
+	 */
+	public void initialize(JButton goSource, JButton haltSource, JButton clearSource, JSlider s1, JSlider s2, JTextArea programArea, JTextArea messageArea, SlingPanel plotPanel) {
+
+		this.goSource = goSource;
+		this.haltSource = haltSource;
+		this.clearSource = clearSource;
+		this.s1 = s1;
+		this.s2 = s2;
+		this.programArea = programArea;
+		this.messageArea = messageArea;
+		this.plotPanel = plotPanel;
+	}
+
+	/*
+	 * Parse a program and return a command.
+	 */
+	protected Command parse(String program) {
+		Tokenizer t = parser.tokenizer();
+		t.setString(program);
+		TokenAssembly ta = new TokenAssembly(t);
+		target = new SlingTarget(s1, s2);
+		ta.setTarget(target);
+		Assembly out = tryMatch(ta);
+		checkResult(program, ta, out);
+		return command(out);
+	}
+
+	/*
+	 * Execcute the command in a separate thread. This allows, in particular,
+	 * halting a large "for" loop.
+	 */
+	public void run() {
+		try {
+			command.execute();
+			setComputing(false);
+			plotPanel.setPlot(target.getRenderables());
+		} catch (Exception e) {
+			String text = e.toString();
+			if (e.getMessage() != null) {
+				text = e.getMessage();
+			}
+			messageArea.append(text + "\n");
+			setComputing(false);
+		}
+	}
+
+	/*
+	 * Sets the state of the IDE to computing or not. Most of the IDE's controls
+	 * are grayed out during computation of a program.
+	 */
+	protected void setComputing(boolean computing) {
+
+		// computing means everything is disabled, except "Halt"
+
+		goSource.setEnabled(!computing);
+		clearSource.setEnabled(!computing);
+		programArea.setEnabled(!computing);
+		messageArea.setEnabled(!computing);
+
+		haltSource.setEnabled(computing);
+	}
+
+	/*
+	 * Any time a slider changes we re-plot.
+	 */
+	public void stateChanged(ChangeEvent e) {
+		go();
+	}
+
+	/*
+	 * Try to recognize a user's program. Handle some track exceptions but pass
+	 * along any other exceptions.
+	 */
+	protected Assembly tryMatch(TokenAssembly ta) {
+		try {
+			return parser.start().bestMatch(ta);
+		} catch (sjm.examples.book.track.TrackException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)